src/optimizer.cpp

```diff
@@ -124,67 +101,75 @@ geometry_msgs::msg::TwistStamped Optimizer::evalControl(
   const geometry_msgs::msg::Twist & robot_speed,
   const nav_msgs::msg::Path & plan, nav2_core::GoalChecker * goal_checker)
 {
-  for (size_t i = 0; i < settings_.iteration_count_; ++i) {
+  state_.pose = robot_pose;
+  state_.speed = robot_speed;
+
+  bool stop_flag = false;
+  costs_.fill(0);
+
+  for (size_t i = 0; i < settings_.iteration_count; ++i) {
     generated_trajectories_ =
-      generateNoisedTrajectories(robot_pose, robot_speed);
-    auto && costs = critic_manager_.evalTrajectoriesScores(
-      state_, generated_trajectories_, plan, robot_pose, goal_checker);
-    updateControlSequence(costs);
+      generateNoisedTrajectories();
+
+
+    auto data =
```

Comment:
This data will be reallocated every iteration `iteration_count` for which most of this information doesn't change. It might be good to have this struct have an actual member `update()` which you can pass it the things that do change between iterations to update so we don't own the construction / copy every iteration, but especially the conversion of the plan to a tensor. Then create data outside of the `for` loop and call `update` in it at the end of each new iteration.

Reply:
it basically old function arguments, tensor now evaluated once